src/hotspot/share/ci/ciEnv.cpp

```diff
@@ -392,23 +393,65 @@ ciInstance* ciEnv::get_or_create_exception(jobject& handle, Symbol* name) {
   return obj == NULL? NULL: get_object(obj)->as_instance();
 }
 
-ciInstance* ciEnv::ArrayIndexOutOfBoundsException_instance() {
+// ------------------------------------------------------------------
+// helper for implicit exception creation
+#ifdef COMPILER2
+ciInstance* ciEnv::create_implicit_exception(Symbol* name, GraphKit* gk) {
+  VM_ENTRY_MARK;
+  InstanceKlass* ik = SystemDictionary::find_instance_klass(name, Handle(), Handle());
+  jobject objh = NULL;
+  if (ik != NULL) {
+    oop obj = ik->allocate_instance(THREAD);
+    if (!HAS_PENDING_EXCEPTION) {
+      Handle handle = Handle(THREAD, obj);
+      java_lang_Throwable::fill_in_stack_trace_of_implicit_exception(handle, gk);
+      // nmethods are no strong roots so we have to create a global JNI handle
+      // for the created exception in order to keep it alive accross GCs.
+      objh = JNIHandles::make_global(handle);
```

Comment:
The backtrace references the java mirrors corresponding to the methods in the backtrace[1]. Thereby the global JNI handle will keep their classloaders alive and prevent classunloading and also unloading of the nmethod being compiled.

[1] https://github.com/openjdk/jdk/blob/7c9868c0b3c9bd3d305e71f91596190813cdccce/src/hotspot/share/classfile/javaClasses.cpp#L2178-L2182